File: 2016/src/day01/task1/Task1.java
package day01.task1;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;

public class Task1 {
   Path path = Path.of("inputs", "day01", "input.txt");
   Direction direction = Direction.North;
   int x, y;

   void task() {
      try {
         var input = Files.readString(this.path).trim().split(", ");
         for (var instruction : input) {
            var turnDirection = instruction.charAt(0);
            var distanceString = instruction.substring(1, instruction.length());
            var distance = Integer.parseInt(distanceString);

            this.direction = switch (turnDirection) {
               case 'R' -> this.direction.turnRight();
               case 'L' -> this.direction.turnLeft();
               default -> throw new RuntimeException("Invalid direction");
            };
            move(distance);
         }

         System.out.println(Math.abs(this.x) + Math.abs(this.y));
      } catch (Exception e) {
         System.err.println(e);
         System.exit(1);
      }
   }

   void move(int distance) {
      switch (this.direction) {
         case North -> this.y -= distance;
         case South -> this.y += distance;
         case East -> this.x += distance;
         case West -> this.x -= distance;
      }
   }

   public static void main(String[] args) {
      var start = Instant.now();

      new Task1().task();

      System.out.println("Done in " + Duration.between(start, Instant.now()).toString().substring(2).toLowerCase());
   }
}

enum Direction {
   North,
   South,
   East,
   West;

   Direction turnRight() {
      return switch (this) {
         case North -> East;
         case South -> West;
         case East -> South;
         case West -> North;
      };
   }

   Direction turnLeft() {
      return switch (this) {
         case North -> West;
         case South -> East;
         case East -> North;
         case West -> South;
      };
   }
}

File: 2017/src/day06/task1.ts
import path from "node:path";

const PATH = path.join("inputs", "day06", "input.txt");

async function task1() {
   let banks = (await Bun.file(PATH).text()).trim().split("\t").map(Number);
   const prevStates = [[...banks]];

   let i = 0;
   while (true) {
      i++;

      let max = Math.max(...banks);
      const maxIdx = banks.findIndex((i) => i === max);

      banks[maxIdx] = 0;

      for (let j = (maxIdx + 1) % banks.length; max > 0; max--, j = (j + 1) % banks.length) {
         banks[j]!++;
      }

      let isOld = prevStates.some((state) => {
         for (let j = 0; j < state.length; j++) {
            if (state[j] !== banks[j]) {
               return false;
            }
         }
         return true;
      });
      if (isOld) break;
      prevStates.push([...banks]);
   }
   console.log(i);
}

console.time("Task Done");
await task1();
console.timeEnd("Task Done");
